drivers/platform/xilinx/xilinx_spi_pl.c

```diff
@@ -0,0 +1,301 @@
+			XSP_CR_LSB_MSB_FIRST_MASK : 0;
+	mask |= XSP_CR_LSB_MSB_FIRST_MASK;
+
+	_update(desc->extra, XSP_CR_OFFSET, val, mask);
+
+	g_spi_desc = *desc;
+}
+
+/* Configure spi */
+static int32_t _xil_spi_init(struct xspi_desc *xdesc)
+{
+	uint32_t val;
+
+	/* Software reset */
+	_write(xdesc, XSP_SRR_OFFSET, XSP_SRR_RESET_MASK);
+	mdelay(1);
+
+	/* Check if success reset */
+	val = _read(xdesc, XSP_SR_OFFSET);
+	if (val != XSP_DEFAULT_SR_VALUE)
+		return -ENODEV;
+
+	/* Disable interrupts. Drivers only works with pooled mode */
+	_write(xdesc, XSP_IIER_OFFSET, 0);
+
+	/* Deasert chip selects */
+	_write(xdesc, XSP_SSR_OFFSET, 0xFFFFFFFF);
+
+	/* Configure spi device */
+	val = XSP_CR_ENABLE_MASK |
+		/* Maybe not needed */
+		//XSP_CR_MANUAL_SS_MASK |
+		XSP_CR_MASTER_MODE_MASK |
+		XSP_CR_TRANS_INHIBIT_MASK;
+
+	_write(xdesc, XSP_CR_OFFSET, val);
+
+	return SUCCESS;
+}
+
+static int32_t xil_spi_init_pl(struct spi_desc **desc,
+	const struct spi_init_param *param)
+{
+	struct spi_desc		*ldesc;
+	struct xspi_desc	*xdesc;
+	int32_t			err;
+	XSpi_Config		*cfg;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	err = -EINVAL;
+	cfg = XSpi_LookupConfig(param->device_id);
+	if (!cfg)
+		goto error;
+
+	if (param->chip_select > cfg->NumSlaveBits)
+		return -EINVAL;
+
+	err = -ENOMEM;
+	ldesc = (struct spi_desc *)calloc(1, sizeof(*ldesc));
+	if (!ldesc)
+		goto error;
+
+	xdesc = (struct xspi_desc *)calloc(1, sizeof(*xdesc));
+	if (!xdesc)
+		goto error;
+
+	xdesc->base_addr = cfg->BaseAddress;
+	xdesc->disable_slaves_mask = (1 << cfg->NumSlaveBits) - 1;
+
+	err = _xil_spi_init(xdesc);
+	if (IS_ERR_VALUE(err))
+		goto error;
+
+	ldesc->extra = xdesc;
+	*desc = ldesc;
+
+	return SUCCESS;
+
+error:
+	free(xdesc);
+	free(ldesc);
+
+	return err;
+}
+
+static int32_t xil_spi_remove_pl(struct spi_desc *desc)
+{
+	if (!desc)
+		return -EINVAL;
+
+	free(desc->extra);
+	free(desc);
+
+	return SUCCESS;
+}
+
+/* Fill fifo until fifo is full or no more data available */
+static inline void _wr_fifo(struct xspi_desc *xdesc, uint8_t *data,
+		     uint32_t *remaining)
+{
+	uint32_t i;
+
+	i = _read(xdesc, XSP_TFO_OFFSET);
+	while (i < FIFO_DEPTH && (*remaining)) {
+		_write(xdesc, XSP_DTR_OFFSET, data[i]);
+		(*remaining)--;
+		i++;
+	}
+}
+
+/* Read fifo until fifo is empty or no more buffer available to read into */
+static inline void _rd_fifo(struct xspi_desc *xdesc, uint8_t *data,
+		     uint32_t *remaining)
+{
+	uint32_t i;
+	uint32_t rx_available;
+
+	rx_available = _read(xdesc, XSP_RFO_OFFSET);
+	i = 0;
+	while (i <= rx_available && *remaining) {
+		data[i] = _read(xdesc, XSP_DRR_OFFSET);
+		(*remaining)--;
+		i++;
+	}
+}
+
+/* If assert is 0 deassert CS, else assert */
+static int32_t xil_spi_cs_assert(struct spi_desc *desc, uint8_t assert)
+{
+	struct xspi_desc	*xdesc;
+
+	if (!desc || !desc->extra)
+		return -EINVAL;
+
+	xdesc = desc->extra;
+	if (assert)
+		_write(xdesc, XSP_SSR_OFFSET, ~(1 << desc->chip_select));
```

Comment:
I don't know much about the hardware here but i know that for xilinx PS, the chip select can be used as a mask to control 8 chip selects using 3 bits (CS Decode I think it's called)

The way the implementation is done here, it would assert / deassert all of them which I think it is not what you intend.
I'm not sure if the PL implementation has this CS Decode mode but it'd be worth investigating.

Reply:
From the datasheet: "At most, one bit can be asserted Low."
So my understanding is that CS Decode won't work for this device